File: PinetreeShop/PinetreeShop.Domain.OrderProcess/OrderProcessManager.cs
using PinetreeShop.CQRS.Infrastructure;
using System;
using System.Collections.Generic;
using System.Linq;
using PinetreeShop.Domain.Baskets.Events;
using PinetreeShop.Domain.Products.Commands;
using PinetreeShop.Domain.Products.Events;
using PinetreeShop.Domain.Orders.Commands;
using PinetreeShop.CQRS.Infrastructure.Repositories;
using PinetreeShop.Domain.Orders;
using PinetreeShop.Domain.Orders.Events;
using PinetreeShop.Domain.OrderProcess.Commands;
using PinetreeShop.Domain.Shared.Types;
using PinetreeShop.Domain.Products;

namespace PinetreeShop.Domain.OrderProcess
{
    public class OrderProcessManager : ProcessManagerBase
    {
        private Guid _basketId;
        private Dictionary<Guid, bool> _reservations = new Dictionary<Guid, bool>();
        private Dictionary<Guid, OrderLine> _orderLines = new Dictionary<Guid, OrderLine>();
        private Address _shippingAddress;
        private Guid _orderId;
        private OrderAggregate.OrderState _state;

        public OrderProcessManager()
        {
            RegisterEventHandler<BasketCheckedOut>(Apply);
            RegisterEventHandler<ProductReserved>(Apply);
            RegisterEventHandler<ProductReservationFailed>(Apply);
            RegisterEventHandler<OrderCreated>(Apply);
            RegisterEventHandler<CreateOrderFailed>(Apply);
            RegisterEventHandler<OrderShipped>(Apply);
            RegisterEventHandler<OrderDelivered>(Apply);
            RegisterEventHandler<OrderCancelled>(Apply);
        }

        #region Internal Event handlers

        private void Apply(BasketCheckedOut evt)
        {
            ProcessId = evt.Metadata.ProcessId;
            _orderId = ProcessId;
            _basketId = evt.AggregateId;
            _shippingAddress = evt.ShippingAddress;
            _orderLines = evt.OrderLines.ToDictionary(ol => ol.ProductId, ol => ol);
            
            DispatchCommand<OrderAggregate>(new CreateOrder(_orderId, _basketId, _shippingAddress));
        }

        private void Apply(ProductReserved evt)
        {
            // todo: if order not pending revoke reservation
            if(_state != OrderAggregate.OrderState.Pending)
            {
                DispatchCommand<ProductAggregate>(new CancelProductReservation(evt.AggregateId, evt.QuantityToReserve));
                return;
            }

            _reservations[evt.AggregateId] = true;

            
            var orderLine = _orderLines[evt.AggregateId];
            DispatchCommand<OrderAggregate>(new AddOrderLine(_orderId, orderLine));

            if(_reservations.All(x => x.Value))
            {
                DispatchCommand<OrderAggregate>(new PrepareOrderForShipping(_orderId));
            }
        }

        private void Apply(ProductReservationFailed obj)
        {
            DispatchCommand<DummyNotifier>(new NotifyAdmin(AggregateRepository.CreateGuid()));
        }

        private void Apply(OrderCreated evt)
        {
            _orderId = evt.AggregateId;
            _state = OrderAggregate.OrderState.Pending;

            foreach (var orderLine in _orderLines.Values)
            {
                _reservations[orderLine.ProductId] = false;
                DispatchCommand<ProductAggregate>(new ReserveProduct(orderLine.ProductId, orderLine.Quantity));
            }
        }

        private void Apply(CreateOrderFailed obj)
        {
            DispatchCommand<DummyNotifier>(new NotifyAdmin(AggregateRepository.CreateGuid()));
        }

        private void Apply(OrderCancelled obj)
        {
            _state = OrderAggregate.OrderState.Cancelled;

            var reservedOrders = _orderLines.Values.Where(ol => _reservations.Any(r => r.Key == ol.ProductId && r.Value));
            foreach (var ol in reservedOrders)
            {
                DispatchCommand<ProductAggregate>(new CancelProductReservation(ol.ProductId, ol.Quantity));
            }
            DispatchCommand<DummyNotifier>(new NotifyCustomer(AggregateRepository.CreateGuid()));
        }

        private void Apply(OrderShipped evt)
        {
            _state = OrderAggregate.OrderState.Shipped;
            var reservedOrders = _orderLines.Values.Where(ol => _reservations.Any(r => r.Key == ol.ProductId && r.Value));
            foreach (var ol in reservedOrders)
            {
                DispatchCommand<ProductAggregate>(new PurchaseReservedProduct(ol.ProductId, ol.Quantity));
            }
            DispatchCommand<DummyNotifier>(new NotifyCustomer(AggregateRepository.CreateGuid()));
        }

        private void Apply(OrderDelivered obj)
        {
            _state = OrderAggregate.OrderState.Delivered;
            DispatchCommand<DummyNotifier>(new NotifyAdmin(AggregateRepository.CreateGuid()));
        }

        #endregion
    }
}

File: PinetreeShop/PinetreeShop.Domain.OrderProcess/OrderProcessManager.fs
module PinetreeShop.Domain.OrderProcess.OrderProcessManager

module Order = PinetreeShop.Domain.Orders.OrderAggregate
module Basket = PinetreeShop.Domain.Baskets.BasketAggregate
module Product = PinetreeShop.Domain.Products.ProductAggregate

open PinetreeCQRS.Infrastructure.Types
open PinetreeCQRS.Infrastructure.Events
open Chessie.ErrorHandling
open System

type ProcessError = 
    | ProcessError of string
    interface IError
    override e.ToString() = sprintf "%A" e

module private Handlers = 
    type ProcessState = 
        | NotCreated
        | BasketCheckedOut
        | OrderPending
        | AllItemsAdded
        | ItemReservationFailed
        | OrderCancelled
        | OrderReadyForShipping
        | OrderShipped
        | OrderDelivered
    
    type Reserved = bool
    
    type State = 
        { OrderLines : Map<ProductId, OrderLine>
          Reserved : ProductId list
          ProcessState : ProcessState
          OrderId : OrderId }
        static member Zero = 
            { OrderLines = Map.empty
              Reserved = []
              ProcessState = NotCreated
              OrderId = OrderId Guid.Empty }
    
    let itemToOrderLine (item : BasketItem) = 
        { ProductId = item.ProductId
          ProductName = item.ProductName
          Price = item.Price
          Quantity = item.Quantity }
    
    let applyEvent state (event : EventEnvelope<IEvent>) = 
        match event.Payload with
        | :? Basket.Event as be -> 
            match be with
            | Basket.BasketCheckedOut(_, items) -> 
                { state with OrderLines = 
                                 List.map (fun (i : BasketItem) -> i.ProductId, (itemToOrderLine i)) items |> Map.ofList }
            | _ -> state
        | :? Product.Event as pe -> 
            match pe with
            | Product.ProductReserved qty -> 
                let (AggregateId productGuid) = event.AggregateId
                let productId = ProductId productGuid
                { state with Reserved = productId :: state.Reserved }
            | _ -> state
        | :? Order.Event as oe -> 
            match oe with
            | Order.OrderCreated _ -> 
                { state with ProcessState = OrderPending
                             OrderId = OrderId.FromAggregateId(event.AggregateId) }
            | Order.OrderReadyForShipping -> { state with ProcessState = OrderReadyForShipping }
            | Order.OrderCancelled -> { state with ProcessState = OrderCancelled }
            | Order.OrderShipped -> { state with ProcessState = OrderShipped }
            | Order.OrderDelivered -> { state with ProcessState = OrderDelivered }
            | _ -> state
        | _ -> state
        
    let createCommand queueName aggregateId processId (event : EventEnvelope<IEvent>) command = 
        let cmd = 
            { AggregateId = aggregateId
              CommandId = Guid.NewGuid() |> CommandId
              Payload = command
              ProcessId = Some(processId)
              CausationId = event.CausationId
              CorrelationId = event.CorrelationId
              ExpectedVersion = AggregateVersion.Irrelevant }
        (queueName, cmd)
    
    let processEvent state (event : EventEnvelope<IEvent>) : Result<(QueueName * CommandEnvelope<ICommand>) list, IError> = 
        match event.ProcessId with
        | Some pid -> 
            match event.Payload with
            | :? Basket.Event as be -> 
                match be with
                | Basket.BasketCheckedOut(address, items) -> 
                    let orderLines = List.map itemToOrderLine items
                    let command = 
                        Order.Create(BasketId.FromAggregateId(event.AggregateId), address, orderLines) :> ICommand 
                        |> createCommand Order.orderQueueName (Guid.NewGuid() |> AggregateId) pid event
                    ok [ command ]
                | _ -> ok []
            | :? Product.Event as pe -> 
                match pe with
                | Product.ProductReserved qty -> 
                    let productId = ProductId.FromAggregateId(event.AggregateId)
                    let (OrderId orderGuid) = state.OrderId
                    let command = 
                        Order.ReserveOrderLineProduct productId :> ICommand 
                        |> createCommand Order.orderQueueName (AggregateId orderGuid) pid event
                    ok [ command ]
                | _ -> ok []
            | :? Order.Event as oe -> 
                match oe with
                | Order.OrderCreated (_,_,orderLines) -> 
                    orderLines
                    |> List.map 
                           (fun ol -> 
                           let (ProductId productGuid) = ol.ProductId
                           Product.Reserve(ol.Quantity) :> ICommand 
                           |> createCommand Product.productQueueName (AggregateId productGuid) pid event)
                    |> ok
                | Order.OrderCancelled -> 
                    state.Reserved
                    |> List.map 
                           (fun i -> 
                           let ol = state.OrderLines.[i]
                           let (ProductId productGuid) = ol.ProductId
                           Product.CancelReservation(ol.Quantity) :> ICommand 
                           |> createCommand Product.productQueueName (AggregateId productGuid) pid event)
                    |> ok
                | Order.OrderShipped -> 
                    state.Reserved
                    |> List.map 
                           (fun i -> 
                           let ol = state.OrderLines.[i]
                           let (ProductId productGuid) = ol.ProductId
                           Product.PurchaseReserved(ol.Quantity) :> ICommand 
                           |> createCommand Product.productQueueName (AggregateId productGuid) pid event)
                    |> ok
                | _ -> ok []
            | _ -> ok []
        | _ -> ok []

let makeOrderProcessHandler = 
    makeEventProcessor { Zero = Handlers.State.Zero
                         ApplyEvent = Handlers.applyEvent
                         ProcessEvent = Handlers.processEvent }
